xbmc/XBDateTime.cpp

```diff
@@ -1544,63 +1393,97 @@ std::string CDateTime::GetAsLocalizedTime(TIME_FORMAT format, bool withSeconds /
   return GetAsLocalizedTime("", false);
 }
 
-CDateTime CDateTime::GetAsUTCDateTime() const
+CDateTime CDateTime::GetAsLocalDateTime() const
 {
-  CDateTime time(m_time);
-  time += GetTimezoneBias();
-  return time;
+#if !defined(TARGET_WINDOWS)
+  auto zone = date::make_zoned(date::current_zone(), m_time);
+
+  CDateTime dt;
+
+  dt.m_time = date::floor<KODI::TIME::Duration>(zone.get_local_time());
+
+  return dt;
+#else // defined(TARGET_WINDOWS)
+  SYSTEMTIME lst{}, st{};
+
+  st.wYear = GetYear();
+  st.wMonth = GetMonth();
+  st.wDay = GetDay();
+  st.wDayOfWeek = GetDayOfWeek();
+  st.wHour = GetHour();
+  st.wMinute = GetMinute();
+  st.wSecond = GetSecond();
+
+  auto secs = date::floor<std::chrono::seconds>(m_time);
+  auto msecs = date::floor<std::chrono::milliseconds>(m_time);
+
+  auto ms_delta = date::floor<std::chrono::milliseconds>(m_time - secs);
+  auto ktp_delta = date::floor<KODI::TIME::Duration>(m_time - msecs);
+
+  st.wMilliseconds = ms_delta.count();
+
+  if (!SystemTimeToTzSpecificLocalTime(nullptr, &st, &lst))
+  {
+    CDateTime dt;
+    dt.SetValid(false);
+    return dt;
+  }
+
+  CDateTime ret(lst.wYear, lst.wMonth, lst.wDay, lst.wHour, lst.wMinute, lst.wSecond);
+  ret += std::chrono::milliseconds{lst.wMilliseconds};
+  ret += ktp_delta;
```

Comment:
Maybe implement additional  `CDateTime` constructor with milliseconds parameter (if is necessary).

NOTE: no idea what `ktp_delta` is

Reply:
I noticed that `CDateTime` to `CDateTime` and `CDateTime` to `std::chrono::system_clock::time_point` cut all sub-seconds away. Proper approach is to add seconds first and sub-second interval as nanoseconds later. This still fails to compile so I have to dig outdated date documentation!